lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,629 @@
+--@DOC_wibox_layout_grid_spacing_EXAMPLE@
+-- @property [horizontal_|vertical_|]spacing
+
+--- The space between the grid and the surrounding area.
+-- If direction (horizontal|vertical) is not specified, set both directions and get the preferred orientation
+--@DOC_wibox_layout_grid_spacing_EXAMPLE@
+-- @property [horizontal_|vertical_|]padding
+
+--- Controls if the columns/rows are expanded to use all the available width/height.
+-- If direction (horizontal|vertical) is not specified, set both directions and get the preferred orientation
+--@DOC_wibox_layout_grid_expand_EXAMPLE@
+-- @property [horizontal_|vertical_|]expand
+
+--- Controls if the columns/rows have the same size or if the size depend on the content.
+-- If direction (horizontal|vertical) is not specified, set both directions and get the preferred orientation
+--@DOC_wibox_layout_grid_expand_EXAMPLE@
+-- @property [horizontal_|vertical_|]fixed_size
+
+
+
+-- Return the maximum value of a table
+local function max_value(t)
+    local m = 0
+    for _,v in ipairs(t) do
+        if m < v then m = v end
+    end
+    return m
+end
+
+-- Return the sum of the values in the table
+local function sum_values(t)
+    local m = 0
+    for _,v in ipairs(t) do
+        m = m + v
+    end
+    return m
+end
+
+
+--- Find a widget in a widget_table, by matching the coordinates
+-- using the  `row`:`col` coordinates, and the spans `row_span` and `col_span`
+-- @param widgets_table Table of the widgets present in the grid
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @tparam number row_span The number of rows the widget spans (default to 1)
+-- @tparam number col_span The number of columns the widget spans (default to 1)
+-- @treturn table of index of widget_table
+local function find_widgets_at(widgets_table, row, col, row_span, col_span)
+    if not row or row < 1 or not col or col < 1 then return nil end
+    row_span = (row_span and row_span > 0) and row_span or 1
+    col_span = (col_span and col_span > 0) and col_span or 1
+    local ret = {}
+    for index, data in ipairs(widgets_table) do
+        -- If one rectangular widget is on left side of other
+        local test_horizontal = not (row > data.row + data.row_span - 1 or data.row > row + row_span - 1)
+        -- If one rectangular widget is above other
+        local test_vertical   = not (col > data.col + data.col_span - 1 or data.col > col + col_span - 1)
+        if test_horizontal and test_vertical then
+            table.insert(ret, index)
+        end
+    end
+    -- reverse sort for safe removal of indices
+    table.sort(ret, function(a,b) return a>b end)
+    return #ret > 0 and ret or nil
+end
+
+--- Get the number of rows and columns occupied by the widgets in the grid
+-- @treturn number,number the number of rows and columns
```

Comment:
What about keeping the `get_dimension` method then, without creating a `direction` property?